### PR TITLE
Add trace_backward, masked_select_backward, and take_backward as ops

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44505 Batched grad support for: slice, select, diagonal
* #44494 Add some batched gradient tests
* #44422 Register diag_backward, diagonal_backward, infinitetely...gelu_backward as operators
* #44410 Register cummaxmin_backward, cumprod_backward as operators
* **#44408 Add trace_backward, masked_select_backward, and take_backward as ops**

See #44052 for context.

Test Plan:
- `pytest test/test_autograd.py -v`

Differential Revision: [D23605504](https://our.internmc.facebook.com/intern/diff/D23605504)

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -810,6 +810,19 @@ Tensor masked_select_cpu(const Tensor & self, const Tensor & mask) {
   return masked_select_out_cpu(result, self, mask);
 }
 
+Tensor masked_select_backward(const Tensor& grad, const Tensor& input, const Tensor& mask) {
+  // normally broadcasting is handled implicitly, but here, because we call an inplace
```
Comment (albanD) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:814]: I don't think the reason for the inplace here is optimization. We cannot really do it any other way as we need to have 0s for the unused values.

**Review by albanD — APPROVED** [review]
LGTM
Small comment update in case its not accurate.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:814]: Thanks for the catch! I was copying-pasting straight from the original code, but I agree it's not really an optimization. I'll delete the "as an optimization" part of the comment.

Comment (zou3519) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Discussed offline: there actually is an optimization going on (we could have just done `zeros_like(input).masked_scatter(mask, grad)`. I clarified the comment with the example

**Pull request closed without merging**